File: RB101_RB109_small_problems/easy_2/mutation.rb

array1 = %w(Moe Larry Curly Shemp Harpo Chico Groucho Zeppo)
array2 = []
array1.each { |value| array2 << value }
array1.each { |value| value.upcase! if value.start_with?('C', 'S') }
puts array2

=begin

Each element in array1 is appended to array2. array2's elements are now all pointing to the same string objects as array1.
upcase! is called on the elements in array1 that start with 'C' or 'S'. 
array1 will now be ['Moe', 'Larry', 'CURLY', 'SHEMP', 'Harpo', 'CHICO', 'Groucho', 'Zeppo']
array2 will also be ['Moe', 'Larry', 'CURLY', 'SHEMP', 'Harpo', 'CHICO', 'Groucho', 'Zeppo'] since none of array2's elements have been reassigned to point to a different object.


This feature of Ruby can get us in trouble by causing confusion & unclear impacts on elements of collections when we mutate elements within a collection. 
We can avoid this by using reassignment (or just non-destructive methods) when creating elements in a new collection.

=end

File: JS210_fundamentals/objects/mutation.js
const array1 = ['Moe', 'Larry', 'Curly', 'Shemp', 'Harpo', 'Chico', 'Groucho', 'Zeppo'];
const array2 = [];

for (let i = 0; i < array1.length; i += 1) {
  array2.push(array1[i]);
}

for (let i = 0; i < array1.length; i += 1) {
  if (array1[i].startsWith('C')) {
    array1[i] = array1[i].toUpperCase();
  }
}

console.log(array2); // ['Moe', 'Larry', 'Curly', 'Shemp', 'Harpo', 'Chico', 'Groucho', 'Zeppo']

// Further Exploration 

// If an object literal was a part of the array1 elements pushed to array2, then any changes made to the object literal in array1 would be reflected in array2.

const array1 = ['Moe', 'Larry', 'Curly', 'Shemp', 'Harpo', 'Chico', 'Groucho', 'Zeppo'];
const array2 = array1;

for (let i = 0; i < array1.length; i += 1) {
  if (array1[i].startsWith('C')) {
    array1[i] = array1[i].toUpperCase();
  }
}

console.log(array2); // ['Moe', 'Larry', 'Curly', 'Shemp', 'Harpo', 'Chico', 'Groucho', 'Zeppo']
